### PR TITLE
Fix various race condition

### PR DESCRIPTION
**Pull request by fly1028**
This pull request addresses the following race conditions.
It addresses https://github.com/kahing/goofys/issues/392 and https://github.com/kahing/goofys/issues/377 without deviating from the existing lock ordering pattern.

```
panic: runtime error: index out of range

 goroutine 32561744 [running]:
 github.com/kahing/goofys/internal.(*Inode).isDir(...)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/handles.go:151
 github.com/kahing/goofys/internal.(*Inode).findInodeFunc.func1(0x1e46, 0x863300)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/handles.go:151 +0xaa
 sort.Search(0x1e47, 0xc42789edb0, 0x15)
         /usr/local/go/src/sort/search.go:66 +0x58
 github.com/kahing/goofys/internal.(*Inode).insertChildUnlocked(0xc427015050, 0xc4228f5dd0)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/handles.go:240 +0x13b
 github.com/kahing/goofys/internal.(*Goofys).insertInode(0xc42017cd00, 0xc427015050, 0xc4228f5dd0)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/goofys.go:687 +0x53
 github.com/kahing/goofys/internal.(*Goofys).CreateFile(0xc42017cd00, 0xc89020, 0xc42b7be300, 0xc42e7fb340, 0x0, 0x0)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/goofys.go:973 +0x138
 github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc4203920a0, 0xc4204be180, 0xc89020, 0xc42b7be300, 0x8ba240, 0xc42e7fb340)
         /home/khc/Code/go/src/github.com/jacobsa/fuse/fuseutil/file_system.go:160 +0x547
 created by github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps
         /home/khc/Code/go/src/github.com/jacobsa/fuse/fuseutil/file_system.go:121 +0x176
```

```

 panic: 3750d7c5-d458-427a-90b0-48938e7b421f_staging_instance_29/ftpTest6on.removeName(smallFiles0050746.pdf) but child not found: 23

 goroutine 108 [running]:
 github.com/kahing/goofys/internal.(*Inode).removeChildUnlocked(0xc420210ea0, 0xc4295a3c20)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/handles.go:203 +0x4dd
 github.com/kahing/goofys/internal.(*Inode).removeChild(0xc420210ea0, 0xc4295a3c20)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/handles.go:222 +0x78
 github.com/kahing/goofys/internal.(*Goofys).ForgetInode(0xc4201c4680, 0xc89020, 0xc430bd4f00, 0xc424e10220, 0x8ba340, 0xc424e10220)
         /home/khc/Code/go/src/github.com/kahing/goofys/internal/goofys.go:707 +0x10d
 github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).handleOp(0xc42036e5a0, 0xc4201be3c0, 0xc89020, 0xc430bd4f00, 0x8ba340, 0xc424e10220)
         /home/khc/Code/go/src/github.com/jacobsa/fuse/fuseutil/file_system.go:151 +0x931
 github.com/jacobsa/fuse/fuseutil.(*fileSystemServer).ServeOps(0xc42036e5a0, 0xc4201be3c0)
         /home/khc/Code/go/src/github.com/jacobsa/fuse/fuseutil/file_system.go:119 +0x11c
 github.com/jacobsa/fuse.Mount.func1(0xc830e0, 0xc42036e5a0, 0xc4201be3c0, 0xc42000e0e0)
         /home/khc/Code/go/src/github.com/jacobsa/fuse/mount.go:90 +0x3f
 created by github.com/jacobsa/fuse.Mount
         /home/khc/Code/go/src/github.com/jacobsa/fuse/mount.go:89 +0x4e8
```

### CODE REVIEW

File: internal/goofys.go [hunk 1]
```diff
@@ -975,11 +986,14 @@ func (fs *Goofys) CreateFile(
 
 	inode, fh := parent.Create(op.Name)
 
+	parent.mu.Lock()
+
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-
 	fs.insertInode(parent, inode)
```
Comment (fly1028) [inline internal/goofys.go]: insertInode() has the following comments but we were not acquiring the parent lock before we call.

// LOCKS_REQUIRED(fs.mu)
// LOCKS_REQUIRED(parent.mu)

File: internal/goofys.go [hunk 2]
```diff
@@ -1012,11 +1026,14 @@ func (fs *Goofys) MkDir(
 		return err
 	}
 
+	parent.mu.Lock()
+
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-
 	fs.insertInode(parent, inode)
```
Comment (fly1028) [inline internal/goofys.go]: insertInode() has the following comments but we were not acquiring the parent lock before we call.

// LOCKS_REQUIRED(fs.mu)
// LOCKS_REQUIRED(parent.mu)

File: internal/goofys.go [hunk 3]
```diff
@@ -702,7 +702,18 @@ func (fs *Goofys) ForgetInode(
...
 		if inode.Parent != nil {
-			inode.Parent.removeChild(inode)
+			inode.Parent.mu.Lock()
+			fs.mu.Lock()
+			// get inode and its parent again in case other goroutines removed them before we get the locks
+			inode := fs.inodes[op.Inode]
+			if inode != nil && inode.Parent != nil {
+				parent := fs.inodes[inode.Parent.Id]
+				if parent != nil {
+					parent.removeChildUnlocked(inode)
```
Comment (fly1028) [inline internal/goofys.go]: As shown in the stack trace, there is at least one race condition where we try to remove inode from the parent but it's already removed.

File: internal/goofys.go [hunk 4]
```diff
@@ -692,20 +692,35 @@ func (fs *Goofys) ForgetInode(
...
 		delete(fs.inodes, op.Inode)
 		fs.forgotCnt += 1
-		fs.mu.Unlock()
 
 		if inode.Parent != nil {
-			inode.Parent.removeChild(inode)
+			// check again to make sure parent still exists now that we have both the fs and parent lock. Parent could have been removed by some other operations?
+			parent := fs.inodes[inode.Parent.Id]
+			if parent != nil {
+				parent.removeChildUnlocked(inode)
```
Comment (fly1028) [inline internal/goofys.go]: As shown in the stacktrace, there is at least one race condition where inode is already removed from parent when we try to remove it here. Acquiring both fs and parent locks seems to help.

**Comment by fly1028** [comment]
@kahing , could you review this please?

**Comment by kahing** [comment]
@fly1028 thanks for the contribution! I made a couple changes base on my comment above, could you see if that still solves the problems you saw?

also, in the future if the PR comes with a test that would be even better :-)

**Pull request closed without merging**